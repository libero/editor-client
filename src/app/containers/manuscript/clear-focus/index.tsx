import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import * as manuscriptEditorAction from 'app/actions/manuscript-editor.actions';

export const ClearFocus: React.FC<{ className?: string }> = ({ children, className }) => {
  const dispatch = useDispatch();
  const handleClick = useCallback(() => {
    dispatch(manuscriptEditorAction.removeFocusAction());
  }, [dispatch]);

  return (
    <div className={className} onClick={handleClick}>
      {children}
    </div>
  );
};
