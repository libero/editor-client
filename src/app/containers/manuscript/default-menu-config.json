{
  "button-group": [
    "save",
    "undo",
    "redo",
    "pdf"
  ],
  "menus": [
    "headings",
    "typography",
    "insert"
  ],
  "custom-buttons": []
}