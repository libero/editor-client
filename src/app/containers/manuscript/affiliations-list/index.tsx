import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';
import { useAffiliationStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedAffiliationFormDialog } from 'app/containers/affiliation-form-dialog';
import { ComponentWithId } from 'app/types/utility.types';

export const AffiliationsList: React.FC<ComponentWithId> = ({ id }) => {
  const classes = useAffiliationStyles();
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const editAffiliation = useCallback(
    (aff: Affiliation) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedAffiliationFormDialog,
          props: { affiliation: aff },
          title: 'Edit Affiliation'
        })
      );
    },
    [dispatch]
  );

  const addAffiliation = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedAffiliationFormDialog,
        props: {},
        title: 'Add Affiliation'
      })
    );
  }, [dispatch]);

  const renderAffiliation = (aff: Affiliation): React.ReactNode => (
    <li key={aff.id} className={classes.listItem}>
      <div className={classes.orderLabel}>{affiliations.length > 1 ? `(${aff.label})` : undefined}</div>
      <div className={classes.affiliationInfo}>{aff.getDisplayName()}</div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAffiliation.bind(null, aff)}>
        <EditIcon fontSize="small" />
      </IconButton>
    </li>
  );

  return (
    <section>
      <SectionContainer label="Affiliations" id={id}>
        <ul className={classes.root}>{affiliations.map(renderAffiliation)}</ul>
      </SectionContainer>
      <ActionButton
        title="Add Affiliation"
        variant="addEntity"
        onClick={addAffiliation}
        className={classes.addButton}
      />
    </section>
  );
};
