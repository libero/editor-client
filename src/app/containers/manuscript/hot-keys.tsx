import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { isMacOs } from 'react-device-detect';
import { GlobalHotKeys, configure } from 'react-hotkeys';

import { canApplyMarkToSelection, canRedoChanges, canUndoChanges } from '../../selectors/manuscript-editor.selectors';
import * as manuscriptActions from '../../actions/manuscript.actions';

configure({
  ignoreEventsCondition(event) {
    return false;
  }
});

const withMod = (shortcut: string): string => {
  const mod = isMacOs ? 'cmd' : 'ctrl';
  return `${mod}+${shortcut}`;
};

export const HotKeyBindings: React.FC<{}> = () => {
  const dispatch = useDispatch();

  const canUndo = useSelector(canUndoChanges);
  const canRedo = useSelector(canRedoChanges);
  const canApply = useSelector(canApplyMarkToSelection);

  const invokeUndo = useCallback(() => {
    if (canUndo) {
      dispatch(manuscriptActions.undoAction());
    }
  }, [dispatch, canUndo]);

  const invokeRedo = useCallback(() => {
    if (canRedo) {
      dispatch(manuscriptActions.redoAction());
    }
  }, [dispatch, canRedo]);

  const invokeToggleMark = useCallback(
    (mark: string) => () => {
      if (canApply(mark)) {
        dispatch(manuscriptActions.toggleMarkAction(mark));
      }
    },
    [dispatch, canApply]
  );

  const keymap = {
    UNDO: withMod('z'),
    REDO: [withMod('y'), withMod('shift+z')],
    BOLD: withMod('b'),
    ITALIC: withMod('i'),
    UNDERLINE: withMod('u')
  };

  const handlers = {
    UNDO: invokeUndo,
    REDO: invokeRedo,
    BOLD: invokeToggleMark('bold'),
    ITALIC: invokeToggleMark('italic'),
    UNDERLINE: invokeToggleMark('underline')
  };

  return <GlobalHotKeys allowChanges={true} keyMap={keymap} handlers={handlers} />;
};
