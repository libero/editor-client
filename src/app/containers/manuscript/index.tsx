import React from 'react';
import { useSelector } from 'react-redux';
import { Backdrop, CircularProgress } from '@material-ui/core';

import { isManuscriptLoaded } from '../../selectors/manuscript.selectors';
import './styles.scss';
import { ManuscriptToolbar } from './manuscript-toolbar';
import { ManuscriptEditor } from './manuscript-editor';

const renderBackdrop = (): JSX.Element => (
  <Backdrop open={true}>
    <CircularProgress color="inherit" />
  </Backdrop>
);

export const ManuscriptContainer: React.FC = () => {
  const isLoaded = useSelector(isManuscriptLoaded);
  const renderContent = (): JSX.Element => (
    <React.Fragment>
      <ManuscriptToolbar />
      <ManuscriptEditor />
    </React.Fragment>
  );

  return isLoaded ? renderContent() : renderBackdrop();
};
