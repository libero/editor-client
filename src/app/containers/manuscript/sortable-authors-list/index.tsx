import React, { Ref, useCallback } from 'react';
import { Chip } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useDispatch, useSelector } from 'react-redux';
import { SortableContainer, SortableElement, SortableHandle } from 'react-sortable-hoc';
import { isEqual } from 'lodash';

import DragIcon from '../../../assets/drag-indicator.svg';
import * as manuscriptActions from '../../../actions/manuscript.actions';
import * as manuscriptEditorActions from '../../../actions/manuscript-editor.actions';
import { getAffiliations, getAuthors } from '../../../selectors/manuscript.selectors';
import { useAuthorsListStyles } from './styles';
import { SectionContainer } from '../../../components/section-container';
import { Person } from '../../../models/person';
import { AuthorFormDialog } from '../../author-form-dialog';
import { ActionButton } from '../../../components/action-button';
import { ComponentWithId } from '../../../types/utility.types';

const DragHandle = React.memo(
  SortableHandle(() => <img src={DragIcon} alt="drag handle" aria-hidden={true} className="drag-handle" />),
  isEqual
);

interface ChipRenderComponent {
  className: string;
}

const ChipRenderComponent: React.FC<ChipRenderComponent> = React.memo(
  React.forwardRef((props, ref: Ref<HTMLDivElement>) => {
    return (
      <div className={props.className} ref={ref}>
        <DragHandle />
        {props.children}
      </div>
    );
  }),
  isEqual
);

const SortableItem = SortableElement(({ value, classes, onEdit }) => {
  const affLabels = value.author.getAffiliationsLabels(value.affiliations);

  return (
    <Chip
      classes={{ root: classes.chip, label: classes.chipLabel }}
      label={
        <span>
          {value.author.getDisplayName()}
          <sup>
            {affLabels.length > 0 ? affLabels.join(',') : undefined}
            {value.author.isCorrespondingAuthor ? '*' : undefined}
          </sup>
        </span>
      }
      onDelete={() => onEdit(value.author)}
      component={ChipRenderComponent}
      color="primary"
      deleteIcon={<EditIcon />}
    />
  );
});

const SortableList = React.memo(
  SortableContainer(({ children, className }) => {
    return <div className={className}> {children} </div>;
  }),
  isEqual
);

export const SortableAuthorsList: React.FC<ComponentWithId> = React.memo(({ id }) => {
  const authors = useSelector(getAuthors);
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const classes = useAuthorsListStyles();

  const onSortEnd = useCallback(
    ({ oldIndex, newIndex }) => {
      if (oldIndex !== newIndex) {
        const author = authors[oldIndex];
        dispatch(manuscriptActions.moveAuthorAction({ index: newIndex, author }));
      }
    },
    [authors, dispatch]
  );

  const onAddNewAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  const onEditAuthor = useCallback(
    (author: Person) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: AuthorFormDialog,
          props: { author },
          title: 'Edit Author'
        })
      );
    },
    [dispatch]
  );

  return (
    <section>
      <SectionContainer label="Authors" id={id}>
        <SortableList className={classes.sortableContainer} onSortEnd={onSortEnd} axis="xy" useDragHandle={true}>
          {authors.map((value, index) => (
            <SortableItem
              key={`item-${value.id}`}
              index={index}
              value={{ author: value, affiliations }}
              classes={classes}
              onEdit={onEditAuthor}
            />
          ))}
        </SortableList>
      </SectionContainer>
      <ActionButton
        title="Add Author"
        variant="addEntity"
        onClick={onAddNewAuthor}
        className={classes.addAuthorButton}
      />
    </section>
  );
}, isEqual);
