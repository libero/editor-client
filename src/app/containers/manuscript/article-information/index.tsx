import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useSelector, useDispatch } from 'react-redux';

import { SectionContainer } from 'app/components/section-container';
import { getArticleInformation } from 'app/selectors/manuscript.selectors';
import { useArticleInformationStyles } from 'app/containers/manuscript/article-information/styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ArticleInfoFormDialog } from 'app/containers/article-info-form-dialog';

export const ArticleInformation: React.FC<{}> = () => {
  const articleInfo = useSelector(getArticleInformation);
  const dispatch = useDispatch();
  const editArticleInfo = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  }, [dispatch]);

  const classes = useArticleInformationStyles();
  return (
    <SectionContainer label="Article information" className={classes.sectionContainer}>
      <section className={classes.infoSection}>
        <div>
          <strong> Article DOI: </strong> {articleInfo.articleDOI}
        </div>
        <div>
          <strong> Publisher ID: </strong> {articleInfo.publisherId}
        </div>
      </section>
      <IconButton classes={{ root: classes.editButton }} onClick={editArticleInfo}>
        <EditIcon fontSize="small" />
      </IconButton>
    </SectionContainer>
  );
};
