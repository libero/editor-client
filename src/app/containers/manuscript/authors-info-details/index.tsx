import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import EditIcon from '@material-ui/icons/Edit';
import { isBoolean, isEqual } from 'lodash';
import Interweave from 'interweave';

import { SectionContainer } from 'app/components/section-container';
import { getAuthorAffiliations, getAuthors } from 'app/selectors/manuscript.selectors';
import { Person } from 'app/models/person';
import {
  useAuthorDetailStyles,
  useAuthorsDetailsListStyles
} from 'app/containers/manuscript/authors-info-details/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { stringifyEditorState } from 'app/utils/view.utils';
import { OrcidIcon } from 'app/assets/icons';
import { ComponentWithId } from 'app/types/utility.types';

export const AuthorsInfoDetails: React.FC<ComponentWithId> = ({ id }) => {
  const authors = useSelector(getAuthors);
  const classes = useAuthorsDetailsListStyles();
  const dispatch = useDispatch();

  const addAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  return (
    <section>
      <SectionContainer label="Author Information" id={id}>
        {authors.map((author) => (
          <AuthorInfo key={author.id} author={author} />
        ))}
      </SectionContainer>
      <ActionButton variant="addEntity" title="Add Author" onClick={addAuthor} className={classes.addAuthorButton} />
    </section>
  );
};

interface AuthorInfoProps {
  author: Person;
}

const AuthorInfo: React.FC<AuthorInfoProps> = React.memo(({ author }) => {
  const classes = useAuthorDetailStyles();
  const dispatch = useDispatch();
  const affiliations = useSelector(getAuthorAffiliations)(author);

  const editAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author },
        title: 'Edit Author'
      })
    );
  }, [author, dispatch]);

  return (
    <section className={classes.root}>
      <div className={classes.authorInfoContainer}>
        <div className={classes.authorInfoLine}>
          <strong> {author.getDisplayName()} </strong>
        </div>
        <div className={classes.authorInfoLine}>
          {author.isCorrespondingAuthor ? 'Corresponding Author: ' : undefined}
          {author.email}
        </div>
        <div className={classes.authorInfoLine}>
          {affiliations.map((aff) => (
            <div key={aff.id}>{aff.getDisplayName()}</div>
          ))}
        </div>
        <div className={classes.authorInfoLine}>
          {author.isAuthenticated ? <OrcidIcon className={classes.orcidIcon} /> : undefined}{' '}
          <a href={`https://orcid.org/${author.orcid}`} target="_blank" rel="noopener noreferrer">
            {author.orcid}
          </a>
        </div>
        {isBoolean(author.hasCompetingInterest) ? (
          <div className={classes.authorInfoLine}>
            {author.hasCompetingInterest
              ? `Competing interest: ${author.competingInterestStatement}`
              : 'No Competing interest declared'}
          </div>
        ) : undefined}
        {author.bio ? (
          <div className={classes.authorBio}>
            <Interweave content={stringifyEditorState(author.bio)} />
          </div>
        ) : undefined}
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAuthor}>
        <EditIcon fontSize="small" />
      </IconButton>
    </section>
  );
}, isEqual);
