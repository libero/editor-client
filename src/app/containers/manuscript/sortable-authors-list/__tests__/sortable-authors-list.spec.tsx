import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { SortableAuthorsList } from '../';
import { getInitialHistory, getLoadableStateSuccess } from '../../../../utils/state.utils';
import { EditorState } from 'prosemirror-state';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';
import { moveAuthorAction } from '../../../../actions/manuscript.actions';

jest.mock('@material-ui/core', () => ({
  Chip: ({ label }) => <div data-cmp="chip">{label}</div>,
  Button: () => <div data-cmp="Button"></div>
}));

jest.mock('react-sortable-hoc', () => ({
  SortableContainer: (_) => _,
  SortableElement: (_) => _
}));

describe('Sortable authors list', () => {
  const mockStore = configureMockStore([]);

  it('renders authors list', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      keywordGroups: {},
      authors: [
        {
          id: '4d53e405-5225-4858-a87a-aec902ae50b6',
          firstName: 'Fred',
          lastName: 'Atherden',
          email: 'f.atherden@elifesciences.org',
          orcId: 'https://orcid.org/0000-0002-6048-1470'
        },
        {
          id: 'c3b008e6-4ae9-4ef9-b7cb-854749a1e897',
          firstName: 'Jeanine',
          lastName: 'Smith',
          suffix: 'III'
        }
      ]
    });
    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });
    const wrapper = create(
      <Provider store={store}>
        <SortableAuthorsList />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('dispatches move actions when authors rearranged', () => {
    const authors = [
      {
        id: '4d53e405-5225-4858-a87a-aec902ae50b6',
        firstName: 'Fred',
        lastName: 'Atherden',
        email: 'f.atherden@elifesciences.org',
        orcId: 'https://orcid.org/0000-0002-6048-1470'
      },
      {
        id: 'c3b008e6-4ae9-4ef9-b7cb-854749a1e897',
        firstName: 'Jeanine',
        lastName: 'Smith',
        suffix: 'III'
      }
    ];

    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      keywordGroups: {},
      authors
    });

    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });

    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <SortableAuthorsList />
      </Provider>
    );

    const onSortEnd = wrapper.findWhere((n) => n.prop('onSortEnd')).prop('onSortEnd');
    onSortEnd({ oldIndex: 0, newIndex: 1 });

    expect(store.dispatch).toBeCalledWith(moveAuthorAction({ index: 1, author: authors[0] }));
  });
});
