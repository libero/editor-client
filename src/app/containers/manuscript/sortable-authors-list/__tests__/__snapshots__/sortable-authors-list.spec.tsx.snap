// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sortable authors list renders authors list 1`] = `
<section>
  <div
    className="makeStyles-root-4"
  >
    <label
      className="makeStyles-label-5"
    >
      Authors
    </label>
    <div
      className="makeStyles-sortableContainer-1"
    >
       
      <div
        data-cmp="chip"
      >
        Fred Atherden
      </div>
      <div
        data-cmp="chip"
      >
        Jeanine Smith
      </div>
       
    </div>
  </div>
  <div
    data-cmp="Button"
  />
</section>
`;
